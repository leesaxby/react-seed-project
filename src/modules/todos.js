import { push } from 'react-router-redux';

const ADD_TODO = 'app/todos/ADD_TODO';
const FETCH_TODOS_SUCCESS = 'app/todos/FETCH_TODOS_SUCCESS';
const UPDATE_DONE_STATUS = 'app/todos/UPDATE_DONE_STATUS';
const CHANGE_FILTER = 'app/todos/CHANGE_FILTER';

const addTodo = (todo) => ({
    type: ADD_TODO,
    payload: todo
});

const updateDoneStatus = (_id, doneStatus) => ({
    type: UPDATE_DONE_STATUS,
    payload: { _id, doneStatus }
});

const fetchTodosSuccess = () => ({
    type: FETCH_TODOS_SUCCESS,
    payload: [
        { _id: 1, text: 'Item One', done: false},
        { _id: 2, text: 'Item Two', done: false},
        { _id: 3, text: 'Item Three', done: false}
    ]
});

const fetchTodos = () => dispatch => Promise.resolve().then(() => dispatch(fetchTodosSuccess()));

const changeFilter = (filter) => dispatch => {
    dispatch(push(filter));
    dispatch({
        type: CHANGE_FILTER,
        payload: filter
    });
};

export {
    ADD_TODO,
    FETCH_TODOS_SUCCESS,
    UPDATE_DONE_STATUS,
    addTodo,
    updateDoneStatus,
    fetchTodos,
    changeFilter
};

const initialState = {
    listItems: [],
    filter: 'ACTIVE'
};

const todo = (state, action) => {
    switch (action.type) {
        case ADD_TODO: {
            const ids = state.listItems.map(todo => todo._id);
            const newId = ids.length ? Math.max(...ids) + 1 : 0;
            return Object.assign({}, action.payload, { _id: newId });
        }
        case UPDATE_DONE_STATUS: {
            const listItems = Object.assign(state.listItems);
            const itemIndex = listItems.findIndex(({ _id }) => _id === action.payload._id);
            const newListItem = Object.assign({}, listItems[itemIndex], {
                done: action.payload.doneStatus
            });

            return { [itemIndex]: newListItem };
        }
        default:
            return state;
    }
};

export default function todos(state = initialState, action) {
    switch (action.type) {
        case ADD_TODO: {
            return Object.assign({}, state, {
                listItems: [
                    ...state.listItems,
                    todo(state, action)
                ]
            });
        }
        case FETCH_TODOS_SUCCESS:
            return Object.assign({}, state, {
                listItems: action.payload
            });
        case UPDATE_DONE_STATUS: {
            return Object.assign({}, state, {
                listItems: Object.assign([], state.listItems, todo(state, action))
            });
        }
        case CHANGE_FILTER:
            return Object.assign({}, state, {
                filter: action.payload
            });
        default:
            return state;
    }
}