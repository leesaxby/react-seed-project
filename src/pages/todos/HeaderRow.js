import React from 'react';
import { Row, Col, Image } from 'react-bootstrap';

import listImage from 'Images/list.png';

export default function HeaderRow() {
	return (
	    <Row>
	        <Col sm={1} lg={1} smOffset={4} lgOffset={5}>
	            <Image src={ listImage }
	                   alt="List Image"
	                   style={{ 'height': '200px',
	                            'marginBottom': '20px' }}/>
								asdasdsdasd
	        </Col>
	    </Row>
	);
}
