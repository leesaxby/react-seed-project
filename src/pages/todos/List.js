import React from 'react';
import PropTypes from 'prop-types';
import { List } from '@material-ui/core';
import Item from './list/Item';

export const TodoList = ({ listItems, onItemSelect }) => (
    <List>
        {
            listItems.map((item) => (
                <Item
                    key={item.id}
                    item={item}
                    onItemSelect={onItemSelect} />
            ))
        }
    </List>
);

TodoList.defaultProps = {
    listItems: [],
};

TodoList.propTypes = {
    listItems: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.number.isRequired,
            text: PropTypes.string,
            done: PropTypes.bool.isRequired,
        }),
    ),
    onItemSelect: PropTypes.func.isRequired,
};

export default TodoList;
