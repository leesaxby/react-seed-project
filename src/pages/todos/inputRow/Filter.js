import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl, defineMessages } from 'react-intl';
import { ToggleButtonGroup, ToggleButton } from 'react-bootstrap';

const messages = defineMessages({
    'todo.filter.ariaLabel': {
        id: 'todo.filter.ariaLabel',
        description: 'Label for the todo filter choice',
        defaultMessage: 'choose filter'
    },
    'todo.filter.todo': {
        id: 'todo.filter.todo',
        description: 'Label for the button that filters active todos',
        defaultMessage: 'todo'
    },
    'todo.filter.done': {
        id: 'todo.filter.done',
        description: 'Label for the button that filters complete todos',
        defaultMessage: 'done'
    }
});

export class Filter extends React.Component {

    static propTypes = {
        filter: PropTypes.oneOf([ 'ACTIVE', 'DONE' ]).isRequired,
        onUpdateFilter: PropTypes.func.isRequired,
        intl: PropTypes.shape({
          formatMessage: PropTypes.func.isRequired
        }).isRequired
    };

    render() {
        const { filter, onUpdateFilter, intl } = this.props;
        return(
            <ToggleButtonGroup type="radio"
                               name="filterToggle"
                               role="radiogroup"
                               aria-label={intl.formatMessage(messages['todo.filter.ariaLabel'])}
                               value={filter}
                               onChange={onUpdateFilter}>

                <ToggleButton value="ACTIVE"
                              role="radio"
                              aria-checked={String(filter === 'ACTIVE')}>
                    <FormattedMessage {...messages['todo.filter.todo']} />
                </ToggleButton>

                <ToggleButton value="DONE"
                              role="radio"
                              aria-checked={String(filter === 'DONE')}>
                    <FormattedMessage {...messages['todo.filter.done']} />
                </ToggleButton>

            </ToggleButtonGroup>
        );
    }
}

export default injectIntl(Filter);
