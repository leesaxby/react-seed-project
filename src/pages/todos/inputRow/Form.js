import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup, FormControl } from 'react-bootstrap';
import { injectIntl, intlShape, defineMessages } from 'react-intl';

const messages = defineMessages({
    'todo.form.addItem': {
        id: 'todo.form.addItem',
        description: 'Placeholder for the todo input',
        defaultMessage: 'Add Item'
    }
});

class Form extends React.Component {

    static propTypes = {
        onAddTodoItem: PropTypes.func.isRequired,
        intl: intlShape.isRequired,
    };

    state = {
        newItem: ''
    };

    render() {
        return (
            <form onSubmit={this.submitTodo}>
                <FormGroup bsSize="large">
                    <FormControl type="text"
                                 data-test-id="todo-add-item"
                                 placeholder={this.props.intl.formatMessage(messages['todo.form.addItem'])}
                                 aria-label={this.props.intl.formatMessage(messages['todo.form.addItem'])}
                                 value={this.state.newItem}
                                 onChange={this.updateNewItem}
                                 onKeyPress={this.handleKeyPress}/>
                </FormGroup>
            </form>
        );
    }

    addTodoItem = () => {
        this.props.onAddTodoItem(this.state.newItem);
        this.setState({ newItem: '' });
    }

    updateNewItem = (e) => {
        this.setState({ newItem: e.target.value });
    }

    submitTodo = (e) => {
        this.addTodoItem();
        e.preventDefault();
    }
    
}

export default injectIntl(Form);