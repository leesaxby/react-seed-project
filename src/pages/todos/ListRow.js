import React from 'react';
import List from './listRow/List';

export default function ListRow() {
    return (
        <List />
    );
}
