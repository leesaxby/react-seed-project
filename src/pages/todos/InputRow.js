import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import { Row, Col } from 'react-bootstrap';
import { addTodo } from 'Modules/todos';
import Form from './inputRow/Form';
import Filter from './inputRow/Filter';

export const InputRow = ({ updateFilter, addTodo, filter }) => (
    <Row>
        <Col sm={8} lg={8} smOffset={1} lgOffset={2}>
            <Form onAddItem={ addTodo } />
        </Col>
        <Col sm={3} lg={2}>
            <Filter filter={ filter } onUpdateFilter={ updateFilter } />
        </Col>
    </Row>
);

InputRow.propTypes = {
    filter: PropTypes.oneOf([ 'ACTIVE', 'DONE' ]),
    addTodo: PropTypes.func.isRequired,
    updateFilter: PropTypes.func.isRequired,
};

const mapStateToProps = ({ todos: { filter } }) => ({ filter });

const mapDispatchToProps = (dispatch) => {
    return {
        updateFilter: (filter) => dispatch(push(filter)),
        addTodo: (todo) => dispatch(addTodo({ text: todo, done: false })),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(InputRow);