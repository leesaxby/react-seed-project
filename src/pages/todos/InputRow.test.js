import { addNewTodo } from 'Modules/todos';
import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { push } from 'react-router-redux';
import createComponentWithIntl from 'Test/createComponentWithIntl';

import InputRowContainer, { InputRow } from './InputRow';
import Form from './inputRow/Form';
import Filter from './inputRow/Filter';


describe('An InputRow', () => {
    const setup = () => {
        const props = {
            filter: 'ACTIVE',
            addTodo: jest.fn(),
            updateFilter: jest.fn(),
        };
        const wrapper = shallow(<InputRow {...props} />);

        return {
            props,
            wrapper,
        };
    };

    it('renders component', () => {
        const { props } = setup();
        const tree = createComponentWithIntl(<InputRow {...props} />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('triggers a callback when the Form requests it', () => {
        const { props, wrapper } = setup();

        const form = wrapper.find(Form).first();
        form.prop('onAddItem')();

        expect(props.addTodo.mock.calls).toHaveLength(1);
    });

    it('triggers a callback when the Filter requests it', () => {
        const { props, wrapper } = setup();

        const filter = wrapper.find(Filter).first();
        filter.prop('onUpdateFilter')();

        expect(props.updateFilter.mock.calls).toHaveLength(1);
    });
});

describe('An InputRow Container', () => {
    const setup = () => {
        const initialState = {
            todos: {
                filter: 'ACTIVE',
            },
        };
        const store = configureStore([thunk])(initialState);
        const wrapper = shallow(<InputRowContainer store={store} />);

        return {
            initialState,
            store,
            wrapper,
        };
    };

    it('renders component', () => {
        const { store } = setup();
        const tree = createComponentWithIntl(<InputRowContainer store={store} />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('maps the right portion of the state to InputRow props', () => {
        const { wrapper, initialState } = setup();

        expect(wrapper.find(InputRow).prop('filter')).toEqual(initialState.todos.filter);
    });

    it('dispatches the right actions from InputRow props', () => {
        const { wrapper, store } = setup();

        expect(store.getActions()).toHaveLength(0);

        const newTodo = 'A New Todo';
        const newFilter = 'DONE';
        wrapper.prop('addTodo')(newTodo);
        wrapper.prop('updateFilter')(newFilter);

        const actions = store.getActions();
        expect(actions).toHaveLength(2);
        expect(actions[0]).toEqual(addNewTodo({ text: newTodo, done: false }));
        expect(actions[1]).toEqual(push(newFilter));
    });
});
