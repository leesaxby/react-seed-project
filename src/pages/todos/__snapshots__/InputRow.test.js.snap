// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An InputRow Container renders component 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        aria-label="Add Item"
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        data-test-id="todo-add-item"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="todoInput"
        >
          Add Item
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="todoInput"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div>
      <span>
        Show Done
      </span>
      <span
        className="MuiSwitch-root"
      >
        <span
          aria-disabled={false}
          aria-label="choose filter"
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-170 MuiSwitch-switchBase MuiSwitch-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <span
              className="MuiSwitch-thumb"
            />
            <input
              checked={false}
              className="PrivateSwitchBase-input-173 MuiSwitch-input"
              id="todoStateSwitch"
              onChange={[Function]}
              type="checkbox"
            />
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </span>
        <span
          className="MuiSwitch-track"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`An InputRow renders component 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        aria-label="Add Item"
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        data-test-id="todo-add-item"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="todoInput"
        >
          Add Item
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="todoInput"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div>
      <span>
        Show Done
      </span>
      <span
        className="MuiSwitch-root"
      >
        <span
          aria-disabled={false}
          aria-label="choose filter"
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-170 MuiSwitch-switchBase MuiSwitch-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <span
              className="MuiSwitch-thumb"
            />
            <input
              checked={false}
              className="PrivateSwitchBase-input-173 MuiSwitch-input"
              id="todoStateSwitch"
              onChange={[Function]}
              type="checkbox"
            />
          </span>
        </span>
        <span
          className="MuiSwitch-track"
        />
      </span>
    </div>
  </div>
</div>
`;
